### DISCUSSION TITLE
Is it non-determinism that the order of the row/column of the coarsen matrix when using multigrid?

### DISCUSSION THREAD

**Discussion by lrtfm** [firedrakeproject/firedrake#2795] [Firedrake support]
I am currently using multigrid with the coarse problem solved by PCLU, and the PC failed randomly with the error below 
```shell
[ 0] Error reported by MUMPS in numerical factorization phase: INFOG(1)=-9, INFO(2)=36
```
Now it is sure that the non-deterministic results are caused by the pivoting and the renumbering of MUMPS. See the discussion here: https://lists.mcs.anl.gov/pipermail/petsc-users/2023-March/048136.html.

I am still curious if there may be non-determinism in the coarsening/refining process, which will result in a different order of the result matrix when every time I run the code.

**Comment by dham** [comment]
The mesh refinement and hence the prolongation and restriction operators are expected to be deterministic.

**Comment by lrtfm** [comment]
Thanks for your explanation.

Best wishes,
Zongze
